### PR TITLE
Implement new PublishedEndpoints default in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This is the C++ counterpart of #2812.

### CODE REVIEW

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 1]
```diff
@@ -1238,49 +1219,65 @@ ObjectAdapterI::computePublishedEndpoints()
...
+                [](const EndpointIPtr& endpoint) { return !endpoint->isLoopback(); });
+
+            // Retrieve published host
+            string publishedHost = _communicator->getProperties()->getProperty(_name + ".PublishedHost");
+
+            if (!endpointsNoLoopback.empty())
             {
-                vector<EndpointIPtr> endps = _incomingConnectionFactories[i]->endpoint()->expandIfWildcard();
-                for (vector<EndpointIPtr>::const_iterator p = endps.begin(); p != endps.end(); ++p)
+                endpoints = endpointsNoLoopback;
```
Comment (pepone) [inline cpp/src/Ice/ObjectAdapterI.cpp:1251]: Can we move assign here?
```suggestion
                endpoints = std::move(endpointsNoLoopback);
```

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 2]
```diff
@@ -1238,49 +1219,65 @@ ObjectAdapterI::computePublishedEndpoints()
...
+                    if (::find_if(
+                            newEndpoints.begin(),
+                            newEndpoints.end(),
+                            [newEndpoint](const EndpointIPtr& p) { return *newEndpoint == *p; }) == newEndpoints.end())
                     {
-                        endpoints.push_back(*p);
+                        newEndpoints.push_back(newEndpoint);
                     }
                 }
+                endpoints = newEndpoints;
```
Comment (pepone) [inline cpp/src/Ice/ObjectAdapterI.cpp:1278]: move assign

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 3]
```diff
@@ -1238,49 +1219,65 @@ ObjectAdapterI::computePublishedEndpoints()
...
-                    // Check for duplicate endpoints, this might occur if an endpoint with a DNS name
-                    // expands to multiple addresses. In this case, multiple incoming connection
-                    // factories can point to the same published endpoint.
-                    //
-                    if (::find(endpoints.begin(), endpoints.end(), *p) == endpoints.end())
+                    EndpointIPtr newEndpoint = endpoint->withPublishedHost(publishedHost);
+                    if (::find_if(
+                            newEndpoints.begin(),
+                            newEndpoints.end(),
+                            [newEndpoint](const EndpointIPtr& p) { return *newEndpoint == *p; }) == newEndpoints.end())
```
Comment (pepone) [inline cpp/src/Ice/ObjectAdapterI.cpp:1273]: Should we capture newEndpoint by reference?

File: cpp/src/Ice/ios/iAPEndpointI.mm [hunk 4]
```diff
@@ -303,19 +303,23 @@ ICEIAP_API void registerIceIAP(bool loadOnInitialize)
 }
 
 vector<EndpointIPtr>
-IceObjC::iAPEndpointI::expandIfWildcard() const
+IceObjC::iAPEndpointI::expandHost() const
 {
```
Comment (pepone) [inline cpp/src/Ice/ios/iAPEndpointI.mm:307]: Would be nicer to use C++ 17 vector initialization

```
return vector<EndpointIPtr> { ... }
```

File: cpp/src/IceBT/EndpointI.cpp [hunk 5]
```diff
@@ -192,36 +192,24 @@ IceBT::EndpointI::acceptor(const string& adapterName, const std::optional<SSL::S
 }
 
 vector<IceInternal::EndpointIPtr>
-IceBT::EndpointI::expandIfWildcard() const
+IceBT::EndpointI::expandHost() const
 {
+    // Nothing to do here.
```
Comment (pepone) [inline cpp/src/IceBT/EndpointI.cpp:197]: Here too, use `return vector { ... }`

**Review by pepone — APPROVED** [review]

File: cpp/src/Ice/ObjectAdapterI.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterI.cpp:1251]: Fixed as suggested.

File: cpp/src/Ice/ObjectAdapterI.cpp (hunk 3 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterI.cpp:1273]: Updated as suggested.

File: cpp/src/Ice/ios/iAPEndpointI.mm (hunk 4 above)
Comment (bernardnormier) [reply cpp/src/Ice/ios/iAPEndpointI.mm:307]: Fixed as suggested. Also for IceBT and Opaque

File: cpp/src/Ice/ObjectAdapterI.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterI.cpp]: fixed

**Review by externl — APPROVED** [review]
Looks good!

**Pull request merged**